samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -119,24 +122,34 @@ public void run(StreamApplication app) {
         throw new SamzaException("No jobs to run.");
       }
       plan.getJobConfigs().forEach(jobConfig -> {
-          log.info("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
+          log.debug("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
           String processorId = getProcessorId(config);
+          if (StringUtils.isEmpty(processorId)) {
+            throw new SamzaException("processorId has to be defined when running StreamApplication locally.");
+          }
+
           StreamProcessor processor =
-              createStreamProcessor(processorId, jobConfig, app, new LocalStreamProcessorListener(processorId));
+              createStreamProcessor(processorId, jobConfig, app, new LocalStreamProcessorLifeCycleListener(processorId));
           processor.start();
           processors.add(processor);
         });
-      appStatus = ApplicationStatus.Running;
 
       // 4. block until the processors are done or there is a failure
-      awaitComplete();
+      latch.await();
+
+      if (throwable.get() != null) {
+        appStatus = ApplicationStatus.UnsuccessfulFinish;
+        throw throwable.get();
```

Comment:
I refactored it so now I am setting the appstatus in the callback directly. Please take a look again.